### ISSUE TITLE
How does "include Cysteines" option affect the prediction results?

### ISSUE THREAD

**Issue by bzhousd** [Kuhlman-Lab/ThermoMPNN#9]
First,  thank you for developing this useful tool.   and I have a question about 'include cysteines' option.   the prediction results are different after running  the same prediction with cysteins and without cysteins.  I used [Colab notebook](https://colab.research.google.com/drive/1OcT4eYwzxUFNlHNPk9_5uvxGNMVg3CFA#scrollTo=i06A5VI142NT).  and I used the example pdb 1PGA

<img width="932" alt="withou_c" src="https://github.com/Kuhlman-Lab/ThermoMPNN/assets/48264534/04ef24ef-a98c-4d3e-86e2-034822cffa60">
<img width="960" alt="with_c" src="https://github.com/Kuhlman-Lab/ThermoMPNN/assets/48264534/f9dea4e0-dac2-4c46-b2ee-bd1cc7c6ebf0">

**Comment by hdieckhaus** [comment]
Hello,

Thank you for bringing this to our attention. This option should not impact the results for the remaining residues, so this is a bug. I have isolated the issue and updated the repository with a quick fix. Please give it a try and let me know if this resolves the issue. Make sure to re-load the Colab session so it can download the new repo, with Runtime -> Restart session.

If you are interested to know the root cause, read on.

It turned out to be a combination of two bugs, both of which added a small amount of noise to the input features during inference. This is usually done during training for regularization purposes, but was mistakenly left on during inference. As a result, even though the random seed was set for inference, passing different mutation sets (with/without cysteine) led to different results. Due to this same bug, running inference multiple times on the same mutation sets could also produce slightly different results. This should not affect any of the inference scripts in the GitHub repository, but since the Colab inference was written separately, it was vulnerable to this.

Hope this helps!

**Comment by bzhousd** [comment]
thank you for quick reply.  I confirm that the results are the same now for both (with/without cysteine).  you said there were two bugs,  I only found one modification, i.e. the augment on structure during inference. could you share the 2nd one?  I converted colab jupter code to local python code,  I guess I need to modify the local python code as well.

Thanks

**Comment by hdieckhaus** [comment]
Hi, sure, not a problem. 

The first modification, as you noted, was adding the augment_eps=0.0 argument to the ProteinMPNN initialization line.

The second modification was adding `model.eval()` before running  `predictions, _ = model(pdb_dict, final_mutation_list)`. This forces the model into eval mode, which disables Dropout layers (previously, these were enabled, leading to random noise being added back into the model during inference). You can find this modification in the Step 5 cell in the Colab notebook.

**Comment by bzhousd** [comment]
Thank you for the quick response.   You are right,  model.eval() is nessary,  dropout lay will take all node instead of randomly pickup some nodes.  it seems that the result is determined,  I tried different seed in torch.manual_seed,  I got identity results.

when I applied thermompnn on my pdb,  I got small ddg on each point mutation ,  I am thinking of a cutoff to pickup the mutation.    do you have suggestion on this cutoff?  the ddg are from -0.8 to 0 in my interested protein area. 

Thanks

**Comment by hdieckhaus** [comment]
Glad to hear it is working! It is a good idea to set a cutoff, since these are ultimately noisy predictions, but picking an exact value is somewhat subjective. We have actually looked into this in the revised paper (should be out shortly), and the general trade-off is as follows: 
- the lower (more negative) you set the cutoff, the more likely you are to get mutations that actually turn out to be stabilizing
- the lower you set the cutoff, the fewer options you're going to have - eventually, you run out of mutations if you go too low

In the literature, people use different cutoffs depending on their system. I would recommend using **-0.5 kcal/mol** as a starting point - mutations below this cutoff had a **success rate of over 50%** in our benchmark study.

I would also recommend manually inspecting the most favored mutations to check for any suspicious results. For example, **check for mutations that might recommend breaking a disulfide or a key hydrogen bond**. At the moment, ThermoMPNN has limited ability to track long side chains, meaning this happens occasionally with lysines etc. (we're working on fixing this). So far, our best results have come from using ThermoMPNN to nominate mutations to improve hydrophobic packing by filling core gaps, so I wouldn't be surprised if many of the favored mutations involve nonpolar-to-nonpolar swaps.
